### PR TITLE
Add AWS x-ray support for API Gateway

### LINKED ISSUE #5564: Add support for AWS x-ray on AWS API Gateway
# This is a Feature Proposal

## Description
Adding support for enabling tracing on an API Gateway would help enrich the information returned when using API Gateway with lambdas leveraging X-ray.  Particularly because when enabling tracing through the CloudFormation template it is done on the API Gateway Deployment object which is not easy to override due to the random string appended to the object name by the serverless framework.  

One particular useful gain with using x-ray with API Gateway is that it resolves issues of x-ray returning successes when API Gateway fails but a lambda responding to the gateway event doesn't.  This happens in instances when the lambda may return a malformed response to the Gateway.

Additionally, as a side note, using X-ray with lambdas and serverless could be done by overriding the function with the appropriate configuration for x-ray.  This could be a work around until x-ray is supported in serverless.

Enabling X-ray for gateway would be done with a `stageDescription` object 
```
stageDescription:
   tracingEnabled: boolean
```
The idea is to allow for extending the `stageDescription` with additional properties in future pull requests.
  
References:
* API Gateway Deployment CFT Spec: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-deployment.html
* API Gateway Stage Description CFT Spec: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apigateway-deployment-stagedescription.html

Similar or dependent issues:
* Support for x-ray with Lambdas #3495, PR #3742 
* Different proposal for adding stage description in stage property: #5400

### PR DESCRIPTION
**Pull request by softprops**
## What did you implement:

Closes #5564

This changes adds support for declaratively enabling [active tracing](https://docs.aws.amazon.com/xray/latest/devguide/xray-services-apigateway.html) for apigateway

## How did you implement it:

If tracing is enabled we'll create a separate `Stage` resource. This feature is opt-in through explicit tracing configuration. When using a `Stage` resource we'll remove the `StageName` from the `Deployment` resource.

We can extend the `Stage` resource configurability later on via a dedicated `serverless.yml` section in `provider`.

## How can we verify it:

To enable this you should only have to add `provider:apiGateway.stageDescription.enableTracing`

```yml
provider:
  name: aws
  tracing:
    apiGateway: true
```

## Todos:

- [X] Write tests
- [X] Write documentation
- [x] Fix linting errors
- [X] Make sure code coverage hasn't dropped
- [X] Provide verification config / commands / resources
- [X] Enable "Allow edits from maintainers" for this PR
- [X] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by softprops** [comment]
while unit tests are passing an integration test im trying is not

serverless.yml 

```yml
service: gw-tracing
provider:
  name: aws
  apiGateway:
    stageDescription:
      tracingEnabled: true
functions:
  hello:
    runtime: python3.6
    handler: hello.handler
    events:
      - http: any /
```

hello.py

```py
def handler(event, ctx):
  return {
    statusCode: 200
  }
```

package.json

```json
{
  "dependencies": {
    "serverless": "softprops/serverless#deployment-stage-description"
  }
}
```

```sh
$ npx serverless deploy
...

  Serverless Error ---------------------------------------

  An error occurred: ApiGatewayDeployment1547359246920 - StageDescription cannot be specified when stage referenced by StageName already exists.
```

I saw a discussion about a similar issue [here](https://forums.aws.amazon.com/thread.jspa?threadID=230706) I'll let you know what I find

File: lib/plugins/aws/provider/awsProvider.js [hunk 1]
```diff
@@ -482,6 +482,34 @@ class AwsProvider {
...
+      const stageDescription = this.serverless.service.provider.apiGateway.stageDescription;
+      if (!_.isPlainObject(stageDescription)) {
+        throw new Error('REST API stage description must be an object');
+      }
+      const tracingEnabled = stageDescription.tracingEnabled;
+      if (tracingEnabled !== undefined) {
+        if (!_.isBoolean(tracingEnabled)) {
+          throw new Error('REST API stage description tracingEnabled must be a boolean');
+        }
+        return {
```
Comment (softprops) [inline lib/plugins/aws/provider/awsProvider.js]: I'm letting `tracingEnabled` be the only supported (and documented) field associated with stage descriptions for now, but having this method in place should make it straightforward and easy to add additional supported fields in the future. Something I've not yet figured out is is there is a better place to communicate structural validation prior to `serverless deploy` invocations.

**Comment by softprops** [comment]
It seems the cloudformation error only happens after the first deploy. I've seen this error references in at least [one other serverless plugin](https://github.com/jacob-meacham/serverless-plugin-bind-deployment-id#known-issues)

**Comment by softprops** [comment]
I'm starting to get the impression stagedescription may indeed not be supported for updates to a stage despite working on the first deploy. I'm thinking another option may be for serverless to somehow manage a stage resource independently https://docs.amazonaws.cn/en_us/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-stage.html

**Comment by eahefnawy** [comment]
@softprops thanks a lot for this PR, and especially for the super helpful comments 🙌 . I'm currently reviewing it.

Is it good to go from your side? or are there any other changes or issues you need to address? Just wanna know the status of this PR.

**Comment by softprops** [comment]
@eahefnawy thanks for posting back.

I think this is going to have to be halted until I figure out the last issue I commented. Help on that would be welcomed. While the code checks according to cloudformation docs out the integration test I posted does not. I need to step a few levels deeper to how cloud formation is actually supposed to behave in this case.

The first deploy works ( with active tracing enabled ) but any deploy afterwards it fails with the error

```
 An error occurred: ApiGatewayDeployment{timestamp} - StageDescription cannot be specified when stage referenced by StageName already exists.
```

I've read through and tried to grok the following similar cases

https://github.com/mapbox/deprecated-hookshot/issues/9#issuecomment-326152641

https://github.com/jacob-meacham/serverless-plugin-bind-deployment-id#known-issues

https://forums.aws.amazon.com/thread.jspa?threadID=230706

This kind of makes me think that in order to get this to work I'd need to change some fundamentals for how serverless manages api gatway deployment stages, potentially as a first class cloud formation resource.

I'd appreciate any help I can get knowledge wise on the history of the current setup for apigw deployments.  Maybe that was tried before but has since switched to its current state. 

The example I provided above should be a good minimal reproducible case for a local test if you have time

**Comment by softprops** [comment]
I can try that route. Another way to look at this to is that it would solve multiple problems when we figure this out. enabling tracing is one problem solved by adding the stage description, here's a list of all the other things that this would enable https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-stage.html

**Comment by jplock** [comment]
Having an explicit StageDescription would help a lot, but probably require people to tear down their existing stacks and redeploy to make use it.

**Comment by eahefnawy** [comment]
@softprops we've discussed it internally, and we'd like to merge it in **only** if you could add the new Stage resource for people activating active tracing in the yml file, otherwise it'd work the same way for the rest of the users.

As @jplock pointed out, this could be a breaking change for some users. Also, if someone is already at the 200 resource limit, they wouldn't be able to make another deployment with this core change.

**Comment by softprops** [comment]
@eahefnawy I still have found a way to work around updates. This only seems to work the first time you deploy. And time you deploy after you'll get the `StageDescription cannot be specified when stage referenced by StageName already exists` error. I'm still not sure what approach would be best to handle that.

You can reproduce the problem with the minimal example posted above

**Comment by AdrieanKhisbe** [comment]
I was working on adding possibility to configure Stage with some`AccessLogSetting` when I come across this issue.
I was developing a plugin `serverless-apigateway-access-logs` that dynamically extend the `
AWS::ApiGateway::Deployment` with a StageDescription. Which works..... only the first time.

> An error occurred: ApiGatewayDeployment1549288734560 - StageDescription cannot be specified when stage referenced by StageName already exists.

I think the best thing to do would be indeed as point out @eahefnawy to have an explicit [`AWS::ApiGateway::Stage`](https://docs.aws.amazon.com/fr_fr/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-stage.html) in serverless we could configure.

I'll be happy to give a hand if needed has this is a current issue we are trying to solve @Coorpacademy

**Comment by softprops** [comment]
@AdrieanKhisbe I [made that suggestion](https://github.com/serverless/serverless/pull/5692#issuecomment-455842888) then @eahefnawy added that we may need to conditionally add this if activating tracing to avoid making this a breaking change. With your use case we'd also have to add another conditional if users opt into configuring the `stage.AccessLogSetting`. 

I'll see if I can get the conditional based on xray tracing soon but id want to maybe see how well that works before adding additional conditional rules for adding a stage resource after.

**Comment by AdrieanKhisbe** [comment]
@softprops I see.

I was just wondering if the explicit `AWS::ApiGateway::Stage` in serverless should be configured via an option, 
option that would be activated by plugins.
This would keep the serverless core simpler and that way, apigateway xray tracing would rather belong to a plugin.

What do you think about this @softprops and @eahefnawy ?

**Comment by softprops** [comment]
I think the hesitation to make the Stage resource default is based on cloudformation limits of 200 resources and some users complain that serverlesses implicit stack should have fewer to opt out of to make more room for application resources.

**Comment by softprops** [comment]
I did a little research on differences and what can be described [Deployment.StageDescription](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apigateway-deployment-stagedescription.html) vs [Stage](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-stage.html?shortFooter=true) resources

Stage Description

```
{
  "AccessLogSetting" : AccessLogSetting,
  "CacheClusterEnabled" : Boolean,
  "CacheClusterSize" : String,
  "CacheDataEncrypted" : Boolean,
  "CacheTtlInSeconds" : Integer,
  "CachingEnabled" : Boolean,
  "CanarySetting" : CanarySetting,
  "ClientCertificateId" : String,
  "DataTraceEnabled" : Boolean,
  "Description" : String,
  "DocumentationVersion" : String,
  "LoggingLevel" : String,
  "MethodSettings" : [ MethodSetting, ... ],
  "MetricsEnabled" : Boolean,
  "Tags" : [ Resource Tag, ... ],
  "ThrottlingBurstLimit" : Integer,
  "ThrottlingRateLimit" : Number,
  "TracingEnabled" : Boolean,
  "Variables" : { String:String, ... }
}
```

Stage

```
{
    "AccessLogSetting" : AccessLogSetting,
    "CacheClusterEnabled" : Boolean,
    "CacheClusterSize" : String,
    "CanarySetting" : CanarySetting,
    "ClientCertificateId" : String,
    "DeploymentId" : String,
    "Description" : String,
    "DocumentationVersion" : String,
    "MethodSettings" : [ MethodSetting, ... ],
    "RestApiId" : String,
    "StageName" : String,
    "Tags" : [ Resource Tag, ... ],
    "TracingEnabled" : Boolean,
    "Variables" : { String:String, ... }
  }
```

Interestingly it seemed like there was more properties in StageDescription than Stage itself
namely `CacheDataEncrypted`, `CacheTtlInSeconds`, `CachingEnabled`, `DataTraceEnabled`,  `LoggingLevel`, `MetricsEnabled`, `ThrottlingBurstLimit`, `ThrottlingRateLimit`

It seems all of these fields are also properties on elements of `Stage.MethodSettings` [MethodSetting](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apigateway-stage-methodsetting.html?shortFooter=true)

This then begs the question if it would be misleading to be exposing stage settings on `provider.apiGateway.stageDescription` when the cloudformation resource serverless would generate, an actual `Stage`, would not be able to represent a direct mapping.

Let me know if this is thinking too far ahead, i.e. supporting stage properties other than tracingEnabled.

**Comment by jplock** [comment]
I wonder if the new v2 API Gateway CloudFormation types can improve this integration at all? https://aws.amazon.com/about-aws/whats-new/2019/02/automate-websocket-api-creation-in-api-gateway-with-cloudformation/

**Comment by exoego** [comment]
FYI:
I was also strugging with `StageDescription` and then switched my strategy to introduce `Stage` in PR #5851.
But found another issue https://github.com/serverless/serverless/pull/5851#issuecomment-465377681.
I will try `APIGatewayV2` as @jplock suggested.

**Comment by softprops** [comment]
@exoego that looks simpler issue I ran into. Let me know if we can collaborate on a solution. I lost a little steam on this after hitting that wall

**Comment by pmuens** [comment]
This looks great! Thanks for working on this @softprops 👍

I just looked over the code and will take it for a spin the upcoming days. @eahefnawy I'll take it from here since I'm also working on x-ray tracing for [Lambda](https://github.com/serverless/serverless/pull/5860).

An important part here would be to figure out how those tracing configs are defined on a provider level. I wrote a proposal [here](https://github.com/serverless/serverless/pull/5860#issuecomment-466385576) but I'm open to other solutions. Main point is that it should be easy and obvious for users to enable tracing for each service individually. Furthermore it would be nice if we could keep the config as generic as possible so that we can follow a pattern when we add additional tracing configs for other services later on.

**Comment by pmuens** [comment]
Hey @softprops thanks again for jumping into this here! This is looking really great 👍

I jumped in today and updated the code slightly so that it matches the interface for the AWS Lambda tracing config (see https://github.com/serverless/serverless/pull/5860). While doing that I simplified the code a little bit since we're currently only focusing on API Gateway tracing via `StageDescription` (we can extend `StageDescription` later on).

During testing I ran into the same issues you've outlined above. Next up I'll look into this...

**Comment by softprops** [comment]
@pmuens yea that's been the main blocker for this pull. See the suggestions above (https://github.com/serverless/serverless/pull/5692#issuecomment-460449186) about creating a Stage resource under specific conditions. I haven't been able to figure out how to do that. Sorry if that wasn't clear here. 

I understand the desire to change the placement of and rename the `StageDescription.TracingEnabled` to `tracing. apiGateway` but a notable tradeoff by moving the position and name to this new serverless framework specific location is that its potentially going to make adding the rest of the stage description fields ( https://github.com/serverless/serverless/pull/5692#issuecomment-460906185 ) less intuitive to author. If Serverless framework ends up exposing other fields like tags, canary settings, metrics enabled ect. Those familiar with the cloudformation  structure might find it odd that the place they do that exists separately from enabling tracing for the stage. 

I'm not sure its a blocker but its a noteworthy design tradeoff for users.

**Comment by pmuens** [comment]
Hey @softprops,
thanks for getting back!

 > @pmuens yea that's been the main blocker for this pull. See the suggestions above ([#5692 (comment)](https://github.com/serverless/serverless/pull/5692#issuecomment-460449186)) about creating a Stage resource under specific conditions. I haven't been able to figure out how to do that. Sorry if that wasn't clear here.

No worries. I just looked through the discussion in this PR and the issue again. The API Gateway implementation is the most compilcated integration we have, so it's sometimes a little bit confusing because there are so many different resources involved... 

> I understand the desire to change the placement of and rename the `StageDescription.TracingEnabled` to `tracing. apiGateway` but a notable tradeoff by moving the position and name to this new serverless framework specific location is that its potentially going to make adding the rest of the stage description fields ( [#5692 (comment)](https://github.com/serverless/serverless/pull/5692#issuecomment-460906185) ) less intuitive to author. If Serverless framework ends up exposing other fields like tags, canary settings, metrics enabled ect. Those familiar with the cloudformation structure might find it odd that the place they do that exists separately from enabling tracing for the stage.

> I'm not sure its a blocker but its a noteworthy design tradeoff for users.

Yes, that's a good point and I totally agree here. IMHO we should expose an easy way to opt-in for tracing via `provider.tracing.apiGateway: true`. This would be a nice DX since it's in-line with `provider.tracing.lambda: true` (this is implemented in the X-Ray tracing for Lambda PR).

In addition to that we should (as you described) expose further configurability via a dedicated stage config in `provider` /cc @exoego. This way we get the best of both worlds (easy to setup tracing, and extensibility in the future).

---

I just worked on another implementation today after reading through the whole thread here. The way I implemented it now is that we'll create a dedicated `Stage` resource if the user enables tracing via `provider.tracing.apiGateway: true`. Additionally we remove the `StageName` from the Deployment resource so that AWS is happy. I tested it today and it's easy to opt-in via a tracing config and opt-out again by removing tracing. Overall this implementation turns this into a non-breaking change and resolves the issue @softprops described above.

It would be nice to get some other 👀 on this PR so that we can ensure that everything works as expected. /cc @serverless/framework-maintainers @eahefnawy @dschep

File: lib/plugins/aws/lib/naming.js [hunk 2]
```diff
@@ -245,6 +245,12 @@ module.exports = {
   getUsagePlanKeyLogicalId(usagePlanKeyNumber) {
     return `ApiGatewayUsagePlanKey${usagePlanKeyNumber}`;
   },
+  getStageLogicalId(stage) {
+    const normalizedStage = this.normalizeName(stage)
+      .replace(/-/g, 'Dash')
+      .replace(/_/g, 'Underscore');
+    return `ApiGatewayStage${normalizedStage}`;
```
Comment (pmuens) [inline lib/plugins/aws/lib/naming.js]: Do we need to include the stage name (e.g. `dev`) here?

Should this be setup that one API can have multiple stages? Otherwise we might want to switch to `ApiGatewayStage` as the logical id...

/cc @eahefnawy @dschep

**Comment by softprops** [comment]
Thanks for helping carry the baton on this. I like where its going.

File: lib/plugins/aws/lib/naming.js (hunk 2 above)
Comment (dschep) [reply lib/plugins/aws/lib/naming.js]: Hmm. Yeah i don't think it's necessary

**Comment by pmuens** [comment]
> Thanks for helping carry the baton on this. I like where its going.

Sure thing! Thanks for starting this @softprops 👍

File: lib/plugins/aws/lib/naming.js (hunk 2 above)
Comment (pmuens) [reply lib/plugins/aws/lib/naming.js]: Sounds reasonable! Thanks @dschep 👍

Any obligations @eahefnawy?

Comment (pmuens) [reply lib/plugins/aws/lib/naming.js]: Just removed this...

**Comment by pmuens** [comment]
Just a quick update here. Today I merged `master` into this PR since x-ray support for Lambda was just merged and having an up-to-date PR makes it easier to test the config for both x-ray tracing capabilities.

After that I removed the suffixing of the stage name in the API Gateway stage resource logical id.

I tested it manually. Creating services from scratch with `tracing: true` is no issue. However enabling tracing on an already existing service results in the `Stage with name "foo" already exists` error. I was definitely not facing this kind of error last week when I tested this exact setup. I'm not sure if it's a race condition. Looking into the CloudFormation console shows me that the new `Deployment` resource is successful deployed before the `Stage` resource is created, which means that the stage should be gone (since the new `Deployment` doesn't include a stage anymore)...

I have to digg deeper here. Any help is greatly appreciated...

**Comment by pmuens** [comment]
I took another deep dive into this today and read through several GitHub issues, blog posts and AWS Forum discussion. There are dozens of proposed workarounds but non actually worked.

After tinkering around with this I stumbled upon a solution which seems to work. Instead of removing the `StageName` from the `Deployment` resource and moving it to the dedicated `Stage` resource I decided to keep it in the `DeploymentResoruce` and just creat a `Stage` resource which references the `Deployment` and has the `TracingEnabled` property set to true. This way AWS is not confused as to what stage it should use / create since it's the same stage name as before.

I tested it with the following setup and tried to break it without success:

```yml
service:
  name: test

provider:
  name: aws
  stage: dev
  runtime: nodejs8.10
  tracing:
    apiGateway: true
    lambda: true

functions:
  hello:
    handler: handler.hello
    events:
      - http: GET hello
  world:
    handler: handler.world
    events:
      - http:
          path: world
          method: POST
```

Here's what I did to test several use cases:

**Enable tracing when service was already deployed in the past **

1. Comment out `tracing.apiGateway`
1. Deploy
1. Bring back `tracing.apiGateway`
1. Deploy
1. Comment out `tracing.apiGateway`
1. Deploy
1. Bring back `tracing.apiGateway`
1. Deploy
1. Remove

**Enable tracing on initial deployment**

1. Ensure that `tracing.apiGateway` is enabled
1. Deploy
1. Comment out `tracing.apiGateway`
1. Deploy
1. Bring back `tracing.apiGateway`
1. Deploy
1. Remove

@eahefnawy @dschep @serverless/framework-maintainers would be nice if you could take a look into this and review it. If there nothing suspicious we should be in a good position to merge and include thi in the upcoming release (together with support for AWS Lambda x-ray tracing).

**Comment by pmuens** [comment]
Ok, so here's another update after I took some time today to test this.

Apparently the last update I made wasn't really solving the issue. CloudFormation was just creating a separate stage with a random name. That's why the deployment had no issues. I since reverted my changes and force-pushed them to this branch.

I did some sophisticated research to actually find a solution for the problem here. Unfortunately I came up empty-handed. There's no way to update an existing deployment without tearing down the whole stack and re-deploying.

The main reason while our current implementation fails is that it creates a new Deployment while keeping the other one alive. After that it tears down the old deployment. Having 2 deployments side-by-side causes the error we see since a stage needs to have a uniqe name. There's no way to tell CloudFormation to tear down the old deployment before deploying the new one (AFAIK this is to reduce downtime).

If anyone stumbled upon this here and has a solution / proposal to fix this, please share it here. I think we did everything we could for now to make this work...

**Comment by pmuens** [comment]
Just looke into [`Amazon API Gateway V2 Resource Types Reference`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-reference-apigatewayv2.html?shortFooter=true) today to see if we could use newer, v2 resources to replace our old v1 `Deployment` and `Stage` resources. The idea here is that `v2` resources might be more robust (orignially brought up by @exoego in https://github.com/serverless/serverless/pull/5851#issuecomment-465377681).

Unfortunately v2 resources are pretty basic and there are no advanced `Stage` configuration parameters such as `TracingEnabled` so we cannot use them as a replacement right now. Maybe AWS will update the functionality over time and we could re-visit this option as a potential solution in the future.

**Comment by orwell1984** [comment]
Hi @pmuens , thanks for your work on this. I was wondering if it's possible to adopt this feature in "create-only" mode, so to speak? The whole CF conundrum is caused by updating existing lambda, right?

**Comment by pmuens** [comment]
Thanks for commenting @orwell1984 :+1:

> Hi @pmuens , thanks for your work on this. I was wondering if it's possible to adopt this feature in "create-only" mode, so to speak? The whole CF conundrum is caused by updating existing lambda, right?

Yes, that sounds like a good plan and seems to be the last resort for the time being to get this merged...

**Comment by pmuens** [comment]
Quick update: I looked into the CloudFormation problems with the current implementation again and it seems like there's no way to resolve them in a non-interruptive fashion.

Given that I've updated the docs and added a comment about the current limitations and the workaround of removing and re-deploying the API Gateway.

Using API Gateway X-Ray Tracing is opt-in, so the user should know what he can expect. Since it's opt-in we should move forward and merge this into `master`.

@dschep @eahefnawy can you take a final look into this and merge if appropriate? Thanks!

**Comment by pmuens** [comment]
I've added some logic which inspects the CloudFormation error message and adds some information on how to solve the problem. This way the user directly sees in the CLI that a remove and re-deployment is necessary.

**Comment by pmuens** [comment]
Here's another update after testing several edge cases with current implementation today:

During testing I experienced an odd behavior. If one enables `tracing`, deploys and afterwards disables `tracing` and deploys again, the stage disappears from the AWS API Gateway console. It can still be found in the compiled CloudFormation template and you can still access the deployed API, but the stage seems to be inaccessible on the AWS console. I'm not sure if this is a problem on AWS end or if it has something to do with our compiled CloudFormation template.

My hunch is that it's because we're going back from a dedicated `AWS::ApiGateway::Stage` resource to an inline `StageName` definition in the `AWS::ApiGateway::Deployment` resource. The odd thing is that we're not seeing any error thrown by AWS.

I'm putting this PR on-hold again since it's not clear if this is a breaking change or if there's anything we can do to work around this issue.

**Comment by pmuens** [comment]
Knock, knock, it's me again...

Just looked into this again. Apparently there's no way to fix this (as we already know). Because of this I implemented a check which will diif the old CloudFormation template with the new one and prints out helpful error messages showing that upgrades require a remove and re-deploy and downgrades might result in unexpected behavior (the user can use `--force` to proceed anyway).

This logic will automatically apply to all future PRs which will also switch to use the new `Stage` resource.

**Comment by jonathanalberghini** [comment]
This is a stupid way to implement.  I updated to 1.41 changed nothing and x-ray was disabled and I get an error and can only deploy with --force.

**Comment by jonathanalberghini** [comment]
And even after deploying once I need to always user the --force option???
provider:
  name: aws
  tracing:
    apiGateway: false

or i get this error
  Serverless Error ---------------------------------------

  NOTE: Disabling API Gateway X-Ray Tracing for existing deployments might result in unexpected behavior.
  We recommend to remove and re-deploy your API Gateway. Use the --force option if you want to proceed with the deployment.

  Please refer to our documentation for more information.

**Comment by gootdude** [comment]
This feature has introduced production blocking bug as well, we can no longer deploy using serverless 1.41.0 as the `checkForBreakingChanges.js` logic always errors and states we must remove our existing API Gateway.

```
          // 1. if the user wants to upgrade to use the new AWS::APIGateway::Stage resource but
          // the old state still uses the stage defined on the AWS::ApiGateway::Deployment resource
          if (oldResources[oldDeploymentLogicalId] && oldResources[oldDeploymentLogicalId].Properties.StageName && newResources[stageLogicalId]) { // eslint-disable-line max-len
            const msg = [
              'NOTE: Enabling API Gateway X-Ray Tracing for existing ',
              'deployments requires a remove and re-deploy of your API Gateway. ',
              '\n\n  ',
              'Please refer to our documentation for more information.',
            ].join('');
            throw new this.serverless.classes.Error(msg);
          }
```
I have provided the relevant portion of my serverless.yml with the cloudformation resources below, we define this to enable detailed access logging in addition to x-ray tracing and set the deployment id using plugin `serverless-plugin-bind-deployment-id`

```
resources:
  Resources:
    ABCApiGateway:
      Type: AWS::ApiGateway::RestApi
      Properties:
        Name: ${self:provider.stage}-ABC
    ApiGatewayStage:
      Type: 'AWS::ApiGateway::Stage'
      Properties:
        DeploymentId:
          Ref: __deployment__
        RestApiId:
          Ref: ABCApiGateway
        StageName: ${self:provider.stage}
        TracingEnabled: true
        MethodSettings:
          - HttpMethod: "*"
            ResourcePath: "/*"
            MetricsEnabled: true
            DataTraceEnabled: false
        AccessLogSetting:
          Format: '{"apiId":"$context.apiId","stage":"$context.stage","resourcePath":"$context.resourcePath","requestId":"$context.requestId","awsEndpointRequestId":"$context.awsEndpointRequestId","xrayTraceId":"$context.xrayTraceId","requestTime":"$context.requestTime","requestTimeEpoch":$context.requestTimeEpoch,"httpMethod":"$context.httpMethod","status":"$context.status","path":"$context.path"}'
          DestinationArn:
                Fn::GetAtt:
                  - ABCAccessLogGroup
                  - Arn
```

**Comment by exoego** [comment]
@jonathannaguin 
@gootdude 
Could you open a new issue with detailed steps and configurations to reproduce?
Thanks in advance !!

**Comment by pmuens** [comment]
> This is a stupid way to implement.
@jonathanalberghini while I agree that this is annoying I have to say that we only had the best intentions when implementing this. The main reason was to provide some safeguards when people want to upgrade and use AWS X-Ray.

Saying that something is "a stupid way to implement" is very harsh and insluting. That's not how you should talk to people on the internet.

---

@exoego thanks for jumping in :+1:

@gootdude thanks for providing in-depth information. I just prepared a PR where I remove the check for breaking changes. Will be out soon and after that I'll publish a patch release.

**Comment by pmuens** [comment]
Quick update that we've just published https://github.com/serverless/serverless/releases/tag/v1.41.1 which should fix the problem.
